### PR TITLE
[FIX] bus, im_livechat: add guest channel instead of using context

### PR DESCRIPTION
**Pull request by tsm-odoo**
In [1], a context object was added to websocket requests to subscribe to guests channels on cors live chats. However, the common mechanism to subscribe to record channels is to pass a string channel that will then be validated by the `_subscribe` method of `ir_websocket`. This approach works with OdooSH out of the box while the other one does not.

This PR removes the mechanism introduce in [1] in favor of the string approach.

[1]: https://github.com/odoo/odoo/pull/130052

### CODE REVIEW

File: addons/mail/models/discuss/ir_websocket.py [hunk 1]
```diff
@@ -22,6 +22,17 @@ def _get_im_status(self, data):
...
+            if isinstance(channel, str) and channel.startswith("mail.guest_"):
+                guest_channels.append(channel)
+                channels.remove(channel)
+        if guest_channels:
+            new_context = {**self.env.context}
+            new_context.update(
+                guest=self.env["mail.guest"]._get_guest_from_token(guest_channels[0].split("_")[1])
+            )
+            self = self.with_context(new_context)
         guest = self.env["mail.guest"]._get_guest_from_context()
```
Comment (seb-odoo) [inline addons/mail/models/discuss/ir_websocket.py:36]: How does it play when there is already a guest in the context? which one should take priority?

File: addons/im_livechat/static/src/embed/common/livechat_service.js [hunk 2]
```diff
@@ -229,17 +229,7 @@ export class LivechatService {
...
-                ...this.busService.context,
-                guest_token: this.guestToken,
-            });
-        }
-        if (this.busService.isActive) {
-            this.busService.forceUpdateChannels();
-        } else {
-            await this.busService.start();
-        }
+        await this.busService.addChannel(`mail.guest_${this.guestToken}`);
```
Comment (seb-odoo) [inline addons/im_livechat/static/src/embed/common/livechat_service.js:232]: Is this awaiting long enough for start to be done before resolving?
As to keep the same behavior as before, not calling init messaging before the bus is ready.

File: addons/im_livechat/static/src/embed/cors/livechat_service_patch.js [hunk 3]
```diff
@@ -0,0 +1,11 @@
+/* @odoo-module */
+
+import { patch } from "@web/core/utils/patch";
+import { LivechatService } from "../common/livechat_service";
+
+patch(LivechatService.prototype, {
+    leaveSession() {
+        this.busService.deleteChannel(`mail.guest_${this.guestToken}`);
```
Comment (seb-odoo) [inline addons/im_livechat/static/src/embed/cors/livechat_service_patch.js:8]: I'm still wondering whether we should delete the channel on leave, or if we should delete it when starting a new session.
Also wondering if we could somehow re-use the same guest if starting a new livechat and the page is still open?
Giving existing guest token to init if possible?

File: addons/im_livechat/static/src/embed/common/livechat_service.js (hunk 2 above)
Comment (tsm-odoo) [reply addons/im_livechat/static/src/embed/common/livechat_service.js:232]: Both `start` and `addChannel` are awaiting `connectionInitializedDeferred` so timing is unchanged.

File: addons/mail/models/discuss/ir_websocket.py (hunk 1 above)
Comment (tsm-odoo) [reply addons/mail/models/discuss/ir_websocket.py:36]: Guest passed explicitly via `channels` is kept.

File: addons/im_livechat/static/src/embed/cors/livechat_service_patch.js (hunk 3 above)
Comment (tsm-odoo) [reply addons/im_livechat/static/src/embed/cors/livechat_service_patch.js]: I think you are right: for non CORS live chats, the guest is kept via the cookie mechanism. I don't see why it should behave differently in the CORS scenario. I adapted the PR to handle this case.

File: addons/im_livechat/controllers/cors/main.py [hunk 4]
```diff
@@ -15,7 +15,11 @@ def cors_visitor_leave_session(self, guest_token, uuid):
     def cors_get_session(
         self, channel_id, anonymous_name, previous_operator_id=None, chatbot_script_id=None, persisted=True, **kwargs
     ):
-        downgrade_to_public_user()
+        guest_token = kwargs.get("guest_token")
+        if guest_token:
+            force_guest_env(guest_token)
+        else:
+            downgrade_to_public_user()
```
Comment (seb-odoo) [inline addons/im_livechat/controllers/cors/main.py:22]: I would not raise on invalid token here, we should be able to create a new session no matter what (obsolete token or whatever).

File: addons/mail/models/discuss/ir_websocket.py (hunk 1 above)
Comment (seb-odoo) [reply addons/mail/models/discuss/ir_websocket.py:36]: Ok but we should keep it only if it is valid then, and probably only the last one given.

**Review by seb-odoo — APPROVED** [review]
@robodoo delegate+

Might be nice to add few tests to cover these a little bit.

**Review by seb-odoo — APPROVED** [review]
@robodoo override=ci/security
Sending on safe channel (guest) and in test file.

**Comment by tsm-odoo** [comment]
@robodoo r+

**Pull request closed without merging**